### ISSUE TITLE
Apple FileProvider support

### ISSUE THREAD

**Issue by JulyIghor** [rclone/rclone#5158]
Hello. Any chance that mac FileProvider feature can be integrated with rclone?
https://developer.apple.com/documentation/fileprovider

**Comment by ncw** [comment]
That looks like lots of Objective C classes which we can't use easily from Go. Is there a public API?

**Comment by JulyIghor** [comment]
> That looks like lots of Objective C classes which we can't use easily from Go. Is there a public API?

maybe this is the way https://coderwall.com/p/l9jr5a/accessing-cocoa-objective-c-from-go-with-cgo

**Comment by footlooseboss** [comment]
API covered at WWDC21: [Video - Sync files to the cloud with FileProvider on macOS](https://developer.apple.com/videos/play/wwdc2021/10182/)

[Create a File Provider extension for macOS](https://developer.apple.com/documentation/fileprovider/macos_support)

[Sample code - Syncing files on macOS](https://developer.apple.com/documentation/fileprovider/macos_support/syncing_files_on_macos)

**Comment by valenokk** [comment]
This feature would be awesome if implemented. 
Fileprovider is just the best option to mount a cloud on macOS.

**Comment by footlooseboss** [comment]
> Fileprovider is just the best option to mount a cloud on macOS.

It's more like native selective sync; it won't replace `rclone mount`. 

From macFUSE developer Benjamin Fleischer:

> File Providers are no file systems. In simple terms, File Providers are a plugin to sync files more or less transparently and store them buried in the the user's home directory. There is no (virtual) volume. macFUSE on the other hand allows you to write full-featured file systems. The File Provider framework lacks many features that macFUSE relies on. Therefore it is not possible to implement macFUSE using File Providers.
> 
> File Providers are tailored to a very specific use case and this is keeping a local copy of your personal files in sync with a cloud storage service.
> 
> File providers do not support common use cases like transparent encryption or decryption. In the File Provider context your files will always be stored unencrypted.
> 
> File Providers don't support streaming file content. So if you want to play a movie file stored in the cloud, the complete file needs to be downloaded first before it can be played. In case of macFUSE, playback starts almost instantly because file content can be streamed.  Appending a couple of bytes to a large file on a macFUSE volume means you will just have to download and upload a couple of bytes of the file. Using File Providers the whole file needs to be download and then uploaded again.
> 
> File Providers are not intended for mounting a volumes in order to access some files real quick and then unmounting the volume again. There is a significant overhead due to File Providers creating local placeholder files for all remote files. In fact, there is no unmount button in Finder for File Providers.
> 
> It would not be possible (or make sense) to build local file systems like NTFS or even network file systems using the File Provider API.
> 
> Dealing with conflicts can be hard using File Providers. In case of macFUSE every write is synchronous. This means you know exactly which range of the file has changed and you can push the changes to your backend (and notify other hosts of the changes) synchronously. This is not supported by File Providers. With File Providers you receive a callback after the file has been changed by the user. You don’t know which part of the file has changed and you have no way of performing any locking while the file is being modified.
> 
> These are just a few differences. There are more.
> 
> TL;DR 
> 
> The macFUSE functionality cannot be moved to the File Provider API. macFUSE is more versatile than File Providers. File System kernel extensions have not been deprecated by Apple, yet. I expect Apple to release a new user space API for developing full-featured file systems. I will do my best to update macFUSE to use this new API instead of the old kernel API as soon as the new API becomes available.

[Future of kexts and File Provider](https://groups.google.com/g/osxfuse-group/c/vY9w1d9N-bQ/m/2h-ICWcuBQAJ)

**Comment by valenokk** [comment]
> File Providers are tailored to a very specific use case and this is keeping a local copy of your personal files in sync with a cloud storage service.

Well, I believe that’s exactly what rclone is trying to achieve, right? 

The current issue with “rclone mount” on MacOS is its complexity. I need to install FUSE. Then I need to create a folder for rclone mount, and them subfolders for each remote I have. Then mount every remote from command line. Then decide which cashing strategy to use for each remote.  Then manually type a path to the aforementioned folder in every app I need to use when I want to access a particular remote. 

File Providers just remove all of those layers of complexity. In ideal world all remotes will just appear in a sidebar in Finder without any additional kernel extensions, like FUSE. And then it’s up to MacOS to decide what to download (when requested) and what to purge when there is not enough free space on SSD. I don’t even see a reason why user will want to unmoumt such a destination, as it will not occupy any extra space. 

Here is what commercial vendors say about File Provider api:
https://www.expandrive.com/strongsync/
Unfortunately this app is of no use to me, as it doesn’t support the clouds I use. 

Rclone on the other side is REALLY flexible. I don’t mind to set it up from terminal (I really like the process, actually), but it’s very tedious to use it from macOS gui after initial set-up. Here is where File Providers api will help enormously.

**Comment by footlooseboss** [comment]
if you want to try Strongsync:

`rclone serve webdav <remote>: --addr :8080 --user rclone --pass rclone`

then add a webdav account http://localhost:8080

If I wanted to stream a 50GB Blu-ray for example I can stream it directly via mount or at that address in a browser. Via File provider I'd need to download it first.

**Comment by footlooseboss** [comment]
[NSFileProviderPartialContentFetching](https://developer.apple.com/documentation/fileprovider/nsfileproviderpartialcontentfetching?changes=latest_minor)

> Support for fetching part of a file’s content.

> Adopt this protocol to let the system request only part of a file. Apps that read files provided by your extension can benefit from this feature, either by minimizing the amount of data your file provider needs to download, or by finishing the download quickly, freeing up the reading process.

> > To trigger a partial download, an app must use POSIX read operations to read part of the file. If you clone the entire file, or read the file using file coordination, the system requests the entire file.

> For example, a photo app could read just the metadata from each picture in a large album, without having to completely download all the images. Alternatively, a video streaming app could begin playing the video before reading the whole file, reading chunks of data just before it needs them.

**Comment by upintheairsheep** [comment]
iSH could be the bridge you are looking for

**Comment by valenokk** [comment]
> iSH could be the bridge you are looking for

I can only find iPhone terminal emulator with this name. Are you sure it's available as macOS app?

**Comment by BasixKOR** [comment]
I researched what would it take to complete this, and I'm not really sure if this is in scope for rclone (although I would love to see this become reality):

- First of all `rclone` should be distributed (or at least distribute) as an `.app` bundle.
  - The app bundle should contain `.appex` bundle which contains another Mach-O binary implementing NSFileProviderReplicatedExtension.
- rclone should have *some* way to be notified when the remote has changed and call `manager.signalEnumerator(for: .workingSet)`; otherwise the system may have a stale data for files that have been opened previously, although this can be noted as a limitation.
- It's techincally possible to write the codes required using [DarwinKit](https://github.com/progrium/darwinkit) as far as I can tell, although it looks like quite a pain to do.

The complexities alone feels like a huge undertaking and I think this could be done better by leveraging [librclone](https://github.com/rclone/rclone/tree/master/librclone) to create a Swift-based application.

**Comment by BasixKOR** [comment]
Unrelated to the comment above but I'm unsure why this issue has a 'new backend' label.
